homeassistant/components/madvr/remote.py

```diff
@@ -0,0 +1,106 @@
+"""Support for MadVR remote control."""
+
+from collections.abc import Iterable
+import logging
+from typing import Any
+
+from madvr.madvr import Madvr
+
+from homeassistant.components.remote import RemoteEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+from .coordinator import MadVRCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the MadVR remote."""
+    coordinator: MadVRCoordinator = entry.runtime_data
+    if not isinstance(coordinator, MadVRCoordinator):
+        raise TypeError("entry.runtime_data is not an instance of MadVRCoordinator")
+
+    madvr_client = coordinator.client
+
+    async_add_entities(
+        [
+            MadvrRemote(hass, coordinator, madvr_client, entry.entry_id),
+        ]
+    )
+
+
+class MadvrRemote(CoordinatorEntity, RemoteEntity):
+    """Remote entity for the MadVR integration."""
+
+    coordinator: MadVRCoordinator
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        coordinator: MadVRCoordinator,
+        madvr_client: Madvr,
+        entry_id: str,
+    ) -> None:
+        """Initialize the remote entity."""
+        super().__init__(coordinator)
+        self.madvr_client = madvr_client
+        self.coordinator = coordinator
+        self._attr_name = coordinator.name
+        self._attr_unique_id = f"{entry_id}_remote"
+        self.entry_id = hass.data[DOMAIN]["entry_id"]
+        self._attr_should_poll = False
+        self.connection_event = self.madvr_client.connection_event
+        self.madvr_client.set_update_callback(coordinator.handle_push_data)
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        await super().async_added_to_hass()
+        # inform the client to start tasks
+        _LOGGER.debug("Adding to hass")
+        _LOGGER.debug("Using loop: %s", self.madvr_client.loop)
+        await self.madvr_client.async_add_tasks()
```

Comment:
So what does this do exactly?

Reply:
this informs the library it should start the background tasks so the integration manages the lifecycle. They will also be cancelled on unload

1. Ping device to determine power state
2. Heartbeat for connection to keepalive
3. Handle commands queue
4. process push notifications from device async

Reply:
Let's do that not in the entity

Reply:
in the coordinator I assume? What are the lifecycle functions for the coordinator I should use?

Reply:
There are none yet

Reply:
okay so how would I move this outside the entity?

Reply:
Create your own function in the coordinator and call that after creating the coordinator